precompiles/gmp/src/lib.rs

```diff
@@ -0,0 +1,312 @@
+
+		// Complete a "Contract Controlled Transfer" with the given Wormhole VAA.
+		// We need to invoke Wormhole's completeTransferWithPayload function, passing it the VAA.
+		// Upon success, it should have transferred tokens to this precompile's address.
+		Self::call(
+			handle,
+			wormhole_bridge,
+			EvmDataWriter::new_with_selector(COMPLETE_TRANSFER_WITH_PAYLOAD_SELECTOR)
+				.write(wormhole_vaa)
+				.build(),
+		)?;
+
+		// query our "after" balance (our being this precompile)
+		let output = Self::call(
+			handle,
+			wrapped_address.into(),
+			EvmDataWriter::new_with_selector(BALANCE_OF_SELECTOR)
+				.write(Address::from(handle.code_address()))
+				.build(),
+		)?;
+		let mut reader = EvmDataReader::new(&output[..]);
+		let after_amount: U256 = reader.read()?;
+		log::debug!(target: "gmp-precompile", "after balance: {}", after_amount);
+
+		let amount_transfered = after_amount.saturating_sub(before_amount);
+
+		// TODO: review
+		if amount_transfered < transfer_with_payload.amount {
+			log::debug!(
+				target: "gmp-precompile",
+				"insufficient funds bridged ({} < {})",
+				amount_transfered, transfer_with_payload.amount);
+			return Err(PrecompileFailure::Revert {
+				exit_status: ExitRevert::Reverted,
+				output: "insufficient funds bridged".into(),
+			});
+		}
+
+		// TODO: Wormhole might have transfered unsupported tokens; we should handle this case
+		//       gracefully (maybe that's as simple as reverting)
+
+		// TODO:
+		let weight_limit: u64 = 1_000_000_000_000u64;
+
+		log::debug!(target: "gmp-precompile", "sending XCM via xtokens::transfer...");
+		let call: orml_xtokens::Call<Runtime> = match user_action {
+			VersionedUserAction::V1(action) => orml_xtokens::Call::<Runtime>::transfer {
+				currency_id,
+				amount,
+				dest: Box::new(VersionedMultiLocation::V3(action.destination)),
+				dest_weight_limit: WeightLimit::Limited(Weight::from_parts(
+					weight_limit,
+					DEFAULT_PROOF_SIZE,
+				)),
+			},
+		};
+
+		log::debug!(target: "gmp-precompile", "sending xcm {:?}", call);
+
+		// TODO: proper origin
+		let origin = Runtime::AddressMapping::into_account_id(handle.code_address());
+		RuntimeHelper::<Runtime>::try_dispatch(handle, Some(origin).into(), call).map_err(|e| {
+			log::debug!(target: "gmp-precompile", "error sending XCM: {:?}", e);
+			e
+		})?;
+
+		Ok(())
+	}
+
+	/// call the given contract / function selector and return its output. Returns Err if the EVM
+	/// exit reason is not Succeed.
+	fn call(
+		handle: &mut impl PrecompileHandle,
+		contract_address: H160,
+		call_data: Vec<u8>,
+	) -> EvmResult<Vec<u8>> {
+		let sub_context = Context {
+			caller: handle.code_address(),
+			address: contract_address,
+			apparent_value: U256::zero(),
+		};
+
+		log::debug!(
+			target: "gmp-precompile",
+			"calling {} ...", contract_address,
+		);
+
+		let (reason, output) = handle.call(
+			contract_address,
+			None,
+			call_data,
+			handle.gas_limit(), // TODO
```

Comment:
`None` already tell the evm to use the remaining gas, so we don't need to worry about it

Reply:
Fixed in 9669e8f.

Also, I think I tracked this behavior down to [this code](https://github.com/rust-blockchain/evm/blob/5fb8a4957c642134782e184c47f7d7b5a29f3629/src/executor/stack/executor.rs#L824), let me know if that's not right.